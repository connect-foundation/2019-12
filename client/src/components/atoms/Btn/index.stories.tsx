import React from 'react';
import { text, boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import Btn from '.';

export default {
  title: 'Atoms / Btn',
};

const styleTypeS = {
  label: 'styleType',
  options: ['primary', 'secondary', 'danger', 'alert', 'success'],
  defaultValue: 'primary',
};

export const general: React.FC = () => (
  <Btn
    styleType={select(
      styleTypeS.label,
      styleTypeS.options,
      styleTypeS.defaultValue,
    )}
    disabled={boolean('disabled', false)}
    content={text('content', 'Button')}
    href={''}
    to={''}
    onClick={action('onClick')}
  />
);

export const anchor: React.FC = () => (
  <Btn
    styleType={select(
      styleTypeS.label,
      styleTypeS.options,
      styleTypeS.defaultValue,
    )}
    disabled={boolean('disabled', false)}
    content={text('content', 'Button')}
    href={text('href', 'https://naver.com')}
    to={''}
    onClick={action('onClick')}
  />
);

export const link: React.FC = () => (
  <>
    <Router>
      <Switch>
        <Route path="*">
          <Btn
            styleType={select(
              styleTypeS.label,
              styleTypeS.options,
              styleTypeS.defaultValue,
            )}
            disabled={boolean('disabled', false)}
            content={text('content', 'Button')}
            href={''}
            to={text('to', '/home')}
            onClick={action('onClick')}
          />
        </Route>
      </Switch>
    </Router>
  </>
);

export const allstyle: React.FC = () => (
  <div
    style={{
      display: 'flex',
      flexDirection: 'column',
    }}
  >
    <Btn
      styleType={'primary'}
      disabled={boolean('disabled', false)}
      content={'Button'}
      href={''}
      to={''}
      onClick={action('onClick')}
    />
    <Btn
      styleType={'secondary'}
      disabled={boolean('disabled', false)}
      content={'Button'}
      href={''}
      to={''}
      onClick={action('onClick')}
    />
    <Btn
      styleType={'danger'}
      disabled={boolean('disabled', false)}
      content={'Button'}
      href={''}
      to={''}
      onClick={action('onClick')}
    />
    <Btn
      styleType={'alert'}
      disabled={boolean('disabled', false)}
      content={'Button'}
      href={''}
      to={''}
      onClick={action('onClick')}
    />
    <Btn
      styleType={'success'}
      disabled={boolean('disabled', false)}
      content={'Button'}
      href={''}
      to={''}
      onClick={action('onClick')}
    />
  </div>
);
