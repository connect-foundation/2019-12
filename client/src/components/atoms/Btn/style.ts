import styled, { css } from 'styled-components';
import { ifProp, palette, theme } from 'styled-tools';
import { Link } from 'react-router-dom';

interface BtnStyleProps {
  styletype: string;
  customProps?: {
    fit: boolean;
    grow: boolean;
  };
}

export const BtnStyle = css<BtnStyleProps>`
  ${theme('fontStyle.button')}
  flex-grow: ${ifProp('customProps.grow', '1', '0')};
  width: ${ifProp('customProps.fit', 'fit-content', '18rem')};
  height: ${ifProp('customProps.fit', 'auto', '4rem')};
  line-height: 1.4rem;
  background-color: ${ifProp('disabled', palette('grayscale', 3), props =>
    palette(props.styletype),
  )};
  flex-shrink: 0;
  cursor: ${ifProp('disabled', 'default', 'pointer')};
  pointer-events: ${ifProp('disabled', 'none', 'auto')};
  color: ${props =>
    props.styletype === 'transparent' ||
    props.styletype === 'transparent-border' ||
    props.styletype === 'transparent-hover'
      ? palette('grayscale', 2)
      : palette('white')};
  -webkit-user-drag: none;
  display: inline-block;
  padding: 1.2rem 2rem;
  border-radius: 0.3rem;
  border-color: ${props =>
    props.styletype === 'transparent-border'
      ? palette('grayscale', 2)
      : 'transparent'};
  outline: none;
  text-align: center;
  user-select: none;
  transition: opacity ease 0.2s, background-color ease 0.2s, color ease 0.2s;
  border-style: solid;
  border-width: thin;

  &:hover {
    opacity: ${palette('opacityscale', 0)};
    ${props => {
      if (props.styletype === 'transparent-border') {
        return css`
          background-color: ${palette('grayscale', 2)};
          color: ${palette('grayscale', 6)};
        `;
      } else if (props.styletype === 'transparent-hover') {
        return css`
          background-color: ${palette('grayscale', 5)};
        `;
      }
    }}
  }
`;

export const Anchor = styled.a`
  ${BtnStyle}
`;

export const StyledBtn = styled.button`
  ${BtnStyle}
  line-height: 0;
`;

export const StyledLink = styled(Link)`
  ${BtnStyle}
`;
