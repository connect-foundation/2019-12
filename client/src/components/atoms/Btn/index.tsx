import React from 'react';
import * as S from './style';

export interface Props {
  /** 버튼 내용 */
  content: string;
  /** react-router/Link 사용 (내부적인 routing)*/
  to?: string;
  /** 외부 링크 */
  href?: string;
  /** button styling type (ex. priamry, secondary) */
  styletype?: string;
  /** disabled 여부 */
  disabled?: boolean;
  /** click handler */
  onClick?: () => void;
  /** flex grow enable */
  grow?: boolean;
  /** fit to size enable */
  fit?: boolean;
}

function Btn({
  content,
  href,
  to,
  styletype = 'primary',
  ...props
}: Props): React.ReactElement {
  if (to) {
    return (
      <S.StyledLink styletype={styletype} to={to} {...props}>
        {content}
      </S.StyledLink>
    );
  }
  if (href) {
    return (
      <S.Anchor styletype={styletype} href={href} {...props}>
        {content}
      </S.Anchor>
    );
  }
  return (
    <S.StyledBtn styletype={styletype} {...props}>
      {content}
    </S.StyledBtn>
  );
}
export default Btn;
