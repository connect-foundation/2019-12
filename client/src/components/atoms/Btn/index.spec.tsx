import React from 'react';
import { mount } from 'enzyme';
import Btn from '.';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

describe('Atom / Btn', () => {
  it('[SNAPSHOT] StyledBtn 렌더링', () => {
    // given
    const wrapper = mount(<Btn styletype={'primary'} children={'Button'} />);

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('[SNAPSHOT] Anchor 렌더링', () => {
    // given
    const wrapper = mount(
      <Btn
        styletype={'primary'}
        children={'Button'}
        href={'https://naver.com'}
      />,
    );

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('[SNAPSHOT] StyledLink 렌더링', () => {
    // react-router-dom's Link 는 Router/Switch/Route 내부에만 위치 가능
    // given
    const wrapper = mount(
      <Router>
        <Switch>
          <Route path="/login">
            <Btn styletype={'primary'} children={'Button'} to={'/home'} />,
          </Route>
        </Switch>
      </Router>,
    );

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('content가 표시된다.', () => {
    // given
    const wrapper = mount(<Btn styletype={'primary'} children={'Button'} />);
    const buttonElement = wrapper.find('button');

    // then
    expect(buttonElement.text()).toBe('Button');
  });

  it('disabled prop에 따라 다른 스타일이 적용된다.', () => {
    // given
    const wrapperIsDisabled = mount(
      <Btn disabled={true} styletype={'primary'} children={'Button'} />,
    );
    const wrapperIsNotDisabled = mount(
      <Btn disabled={false} styletype={'primary'} children={'Button'} />,
    );

    // then
    expect(wrapperIsDisabled.prop('disabled')).toBe(true);
    expect(wrapperIsDisabled).toHaveStyleRule('cursor', 'default');
    expect(wrapperIsDisabled).toHaveStyleRule('pointer-events', 'none');

    expect(wrapperIsNotDisabled.prop('disabled')).toBe(false);
    expect(wrapperIsNotDisabled).toHaveStyleRule('cursor', 'pointer');
    expect(wrapperIsNotDisabled).toHaveStyleRule('pointer-events', 'auto');
  });

  it('fit prop에 따라 다른 스타일이 적용된다.', () => {
    // given
    const wrapperIsFit = mount(
      <Btn fit disabled styletype={'primary'} children={'Button'} />,
    );
    const wrapperIsNotFit = mount(
      <Btn disabled styletype={'primary'} children={'Button'} />,
    );
    // then
    expect(wrapperIsFit).toHaveStyleRule('height', 'auto');
    expect(wrapperIsFit).toHaveStyleRule('width', 'fit-content');
    expect(wrapperIsNotFit).toHaveStyleRule('height', '4rem');
    expect(wrapperIsNotFit).toHaveStyleRule('width', '18rem');
  });
});
