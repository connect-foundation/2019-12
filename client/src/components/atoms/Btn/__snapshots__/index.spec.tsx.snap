// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom / Btn [SNAPSHOT] Anchor 렌더링 1`] = `
.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  width: 18rem;
  height: 4rem;
  line-height: 1.4rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: pointer;
  pointer-events: auto;
  -webkit-user-drag: none;
  display: inline-block;
  padding: 1.2rem 2rem;
  border-radius: 0.3rem;
  border-color: transparent;
  outline: none;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: opacity ease 0.2s,background-color ease 0.2s,color ease 0.2s;
  transition: opacity ease 0.2s,background-color ease 0.2s,color ease 0.2s;
  border-style: solid;
  border-width: thin;
}

<Btn
  href="https://naver.com"
  styletype="primary"
>
  <styled.a
    customProps={
      Object {
        "fit": false,
        "grow": false,
      }
    }
    href="https://naver.com"
    styletype="primary"
  >
    <StyledComponent
      customProps={
        Object {
          "fit": false,
          "grow": false,
        }
      }
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  ",
              "
  ",
              [Function],
              "
  flex-grow: ",
              [Function],
              ";
  width: ",
              [Function],
              ";
  height: ",
              [Function],
              ";
  line-height: 1.4rem;
  background-color: ",
              [Function],
              ";
  flex-shrink: 0;
  cursor: ",
              [Function],
              ";
  pointer-events: ",
              [Function],
              ";
  color: ",
              [Function],
              ";
  -webkit-user-drag: none;
  display: inline-block;
  padding: 1.2rem 2rem;
  border-radius: 0.3rem;
  border-color: ",
              [Function],
              ";
  outline: none;
  text-align: center;
  user-select: none;
  transition: opacity ease 0.2s, background-color ease 0.2s, color ease 0.2s;
  border-style: solid;
  border-width: thin;

  &:hover {
    opacity: ",
              [Function],
              ";
    ",
              [Function],
              "
  }
",
              "
",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      href="https://naver.com"
      styletype="primary"
    >
      <a
        className="c0"
        href="https://naver.com"
      >
        Button
      </a>
    </StyledComponent>
  </styled.a>
</Btn>
`;

exports[`Atom / Btn [SNAPSHOT] StyledBtn 렌더링 1`] = `
.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  width: 18rem;
  height: 4rem;
  line-height: 1.4rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: pointer;
  pointer-events: auto;
  -webkit-user-drag: none;
  display: inline-block;
  padding: 1.2rem 2rem;
  border-radius: 0.3rem;
  border-color: transparent;
  outline: none;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: opacity ease 0.2s,background-color ease 0.2s,color ease 0.2s;
  transition: opacity ease 0.2s,background-color ease 0.2s,color ease 0.2s;
  border-style: solid;
  border-width: thin;
  line-height: 0;
}

<Btn
  styletype="primary"
>
  <styled.button
    customProps={
      Object {
        "fit": false,
        "grow": false,
      }
    }
    styletype="primary"
  >
    <StyledComponent
      customProps={
        Object {
          "fit": false,
          "grow": false,
        }
      }
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  ",
              "
  ",
              [Function],
              "
  flex-grow: ",
              [Function],
              ";
  width: ",
              [Function],
              ";
  height: ",
              [Function],
              ";
  line-height: 1.4rem;
  background-color: ",
              [Function],
              ";
  flex-shrink: 0;
  cursor: ",
              [Function],
              ";
  pointer-events: ",
              [Function],
              ";
  color: ",
              [Function],
              ";
  -webkit-user-drag: none;
  display: inline-block;
  padding: 1.2rem 2rem;
  border-radius: 0.3rem;
  border-color: ",
              [Function],
              ";
  outline: none;
  text-align: center;
  user-select: none;
  transition: opacity ease 0.2s, background-color ease 0.2s, color ease 0.2s;
  border-style: solid;
  border-width: thin;

  &:hover {
    opacity: ",
              [Function],
              ";
    ",
              [Function],
              "
  }
",
              "
  line-height: 0;
",
            ],
          },
          "displayName": "styled.button",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      styletype="primary"
    >
      <button
        className="c0"
      >
        Button
      </button>
    </StyledComponent>
  </styled.button>
</Btn>
`;

exports[`Atom / Btn [SNAPSHOT] StyledLink 렌더링 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch />
  </Router>
</BrowserRouter>
`;
